### PR TITLE
[#15] Add cardano-chain Common modules

### PR DESCRIPTION
**Pull request by ruhatch**
This is the first of a series of PRs to bring the `cardano-chain` package into the codebase. I've split it up, so that hopefully each PR is manageable!

Again this is a first version of this stuff, so you can definitely make it a relatively light review!

### CODE REVIEW

File: cardano-chain.cabal [hunk 1]
```diff
@@ -14,7 +14,167 @@ cabal-version:       >=1.10
 
 library
   hs-source-dirs:      src
-  -- exposed-modules:
-  -- other-modules:
+  exposed-modules:
```
Comment (dnadales) [inline cardano-chain.cabal:17]: I'm a bit scared of the number of modules in this package :D

File: src/Cardano/Chain/Common/AddrAttributes.hs [hunk 2]
```diff
@@ -0,0 +1,95 @@
...
+'Attributes' are conceptually a map, where keys are numbers ('Word8').
+
+For address there are two attributes:
+• 0 — stake distribution, defaults to 'BootstrapEraDistr';
+• 1 — derivation path, defaults to 'Nothing'.
+
+-}
+
+instance Bi (Attributes AddrAttributes) where
+    -- FIXME @avieth it was observed that for a 150kb block, this call to
```
Comment (dnadales) [inline src/Cardano/Chain/Common/AddrAttributes.hs:57]: I'm a bit worried that we're carrying over the FIXME's from the legacy code base. What's our plan regarding these?

File: src/Cardano/Chain/Common/AddrStakeDistribution.hs [hunk 3]
```diff
@@ -0,0 +1,120 @@
...
+    -- ^ Stake distribution for bootstrap era.
+    | SingleKeyDistr !StakeholderId
+    -- ^ Stake distribution stating that all stake should go to the given stakeholder.
+    | UnsafeMultiKeyDistr !(Map StakeholderId CoinPortion)
+    -- ^ Stake distribution which gives stake to multiple
+    -- stakeholders. 'CoinPortion' is a portion of an output (output
+    -- has a value, portion of this value is stake). The constructor
+    -- is unsafe because there are some predicates which must hold:
+    --
+    -- • the sum of portions must be @maxBound@ (basically 1);
```
Comment (dnadales) [inline src/Cardano/Chain/Common/AddrStakeDistribution.hs:42]: Is this haddock compliant?

**Review by dnadales — APPROVED** [review]
Quite a big review :) I looked at it very superficially.

What I'm worried is that we're transferring big portions of the code from `cardano-sl`, which means transferring technical debt as well. Why do we need it? Do we have a plan in place to deal with this?

File: src/Cardano/Chain/Common/AddrAttributes.hs (hunk 2 above)
Comment (ruhatch) [reply src/Cardano/Chain/Common/AddrAttributes.hs:57]: We'll take them on a case by case basis, but there shouldn't be too many. These aren't that old if @avieth wrote this one, so we can get him to provide more context.

**Comment by ruhatch** [comment]
@dnadales We will be carefully reviewing all the types to make sure that we're avoiding technical debt. The `binary`, `crypto`, `core`, and `chain` packages from `cardano-sl` were not the source of most of the issues with the old codebase anyway. I've also already removed a lot of stuff from this import. This is (almost) the bare minimum needed to get the `Bi` golden/round-trip testing of all blockchain types working.

**Pull request merged**